docs/src/man/metadata.md

```diff
@@ -0,0 +1,206 @@
+  call to `metadata` function throws an error. A similar rule applies to calling
+  `hascolmetadata` before calling `colmetadata`.
+* if you know you query a data frame, using the `hasmetadata` function avoids
+  creation of metadata dictionary in a data frame in case it were not needed.
+  A call to `metadata` will create such a dictionary when it was not present
+  in a data frame yet.
+
+## Examples
+
+Here is a simple example how you can work with metadata in DataFrames.jl:
+
+```jldoctest dataframe
+julia> df = DataFame(name=["Jan Krzysztof Duda", "Jan Krzysztof Duda",
+                           "Radosław Wojtaszek", "Radosław Wojtaszek"],
+                     date=["2022-Jun", "2021-Jun", "2022-Jun", "2021-Jun"],
+                     rating=[2750, 2729, 2708, 2687])
+
+julia> hasmetadata(df)
+
+julia> df_meta = metadata(df)
+
+julia> df_meta["name"] = "ELO ratings of chess players"
+
+julia> hasmetadata(df)
+
+julia> metadata(df)
+
+julia> empty!(df_meta)
+
+julia> hasmetadata(df)
+
+julia> metadata(df)
+
+julia> hasmetadata(df, :rating)
+
+julia> rating_meta = metadata(df, :rating)
+
+julia> rating_meta["name"] = "First and last name of a player"
+
+julia> rating_meta["rating"] = "ELO rating in classical time control"
+
+julia> rating_meta["date"] = "Rating date in yyyy-u format for \"english\" locale"
+
+julia> hasmetadata(df, :rating)
+
+julia> metadata(df, :rating)
+
+julia> metadata.(Ref(df), names(df))
+
+julia> empty!(rating_meta)
+
+julia> hasmetadata(df, :rating)
+
+julia> metadata(df, :rating)
+```
+
+As a practical tip if you have metadata attached to some object
+(either data frame or data frame column) and you want to propagate it to
+some new object you can use:
+* `copy!` to fully overwrite destination object metadata with source object
+  metadata;
+* `merge!` to add destination object metadata with source object metadata
+  (overwriting duplicates).
+
+## Propagation of metadata
+
+An important design feature of metatada is how it is handled when you transform
+data frames.
+
+!!! note
+
+    The provided rules might change in the future. Any change to metadata
+    propagation rules will not be considered to be a breaking change
+    in DataFrames.jl and can be done in any minor release of DataFrames.jl.
+    Such changes might be made based on users' feedback about what metadata
+    propagation rules are most convenient in practice.
+
+The general design rules for propagation of table and column level metadata
+are as follows:
+* if some object (data frame or column) is mutated in-place its metadata
+  is not changed.
+* preferably (this is not guaranteed in all cases as sometimes it might be hard
+  to detect that this condition is met) if some object is moved as a whole
+  (either by coping or by aliasing) then its metadata is propagated.
+* in all other cases if a new object is allocated from an existing object and
+  this transformation is not a copy or aliasing then metadata is dropped.
```

Comment:
This means that e.g. `dropmissing` and `transform` wouldn't keep table metadata but `dropmissing!` and `transform!` would? I'd find it weird to make a difference between these. It's quite convenient to think that `f(x) = f!(copy(x))` holds in general.

I agree that if a column is transformed its metadata shouldn't be propagated, but table metadata should probably we kept under most transformations. By definition, changing columns shouldn't make table metadata obsolete as metadata about columns should be stored as columns metadata, so we could propagate table metadata in these cases. Taking a subset of rows could theoretically make table metadata obsolete (e.g. if you store the number of observations) so it's harder to decide, but one could also make a case for preserving it as a lot of information would be useful to propagate (observation year, data collection details, license...).

Regarding column metadata, it seems that we should really preserve it when subsetting rows. Otherwise one would lose e.g. all column labels and units when taking a subset, which makes metadata almost useless. A rule that you mentioned before IIRC would be that if cells values are changed then metadata is lost. A very simple but probably too radical rule would be to consider that metadata is attached to a column name, and assume that as long as you keep the same name, you want to keep the metadata. For example this would mean that if you do `transform!(df, [:income, :household_size] => ((i, s) -> i/sqrt(s)) => outcol)`, then you should use `outcol = :equivalized_income` rather than `outcol = :income`. But this would be misleading if you create a column whose name clashes with an existing one without noticing.

Reply:
OK - I have given it another try :).